bento_wes/runs.py

```diff
@@ -136,89 +148,78 @@ def _create_run(db, c):
-                    f"(Scope: {scope}, OTT URL: {ott_generate_url}, headers included: {list(headers.keys())})")
-
-            one_time_tokens.extend(tr.json())
-
-        # TODO: refactor --
+            one_time_tokens.extend(ott.get(scope, 1))
+
+        # Generate temporary tokens for polling purposes during ingestion
+        # (Gohan specific)
         if chord_mode and tt_endpoint_namespace and "gohan" in workflow_ingestion_url:
-            # TODO: Refactor into class
-            headers = {**auth_header_dict}  # TODO: Host?
-            tt_generate_url = urljoin(tt_endpoint_namespace.rstrip("/") + "/", "generate")
 
-            # Request an additional TT for the service ingest request
             scope = ("/" if chord_url in workflow_ingestion_url else "") + workflow_ingestion_url.replace(
                 chord_url, "").rsplit("/", 1)[0] + "/"
-            tr = requests.post(tt_generate_url, headers=headers, json={
-                # TODO: This sort of assumes the ingest URL is on the same domain as WES, which isn't necessarily
-                #  correct. An error should be thrown if there's a mismatch and we're still trying to do TT stuff
-                "scope": scope,
-                "number": 1,
-            }, verify=not current_app.config["BENTO_DEBUG"])
-
-            if not tr.ok:
-                # An error occurred while requesting TTs, so we cannot complete the run request
-                return flask_internal_server_error(
-                    f"Got error while requesting one-time-use tokens for ingestion URL: {tr.content} "
-                    f"(Scope: {scope}, TT URL: {tt_generate_url}, headers included: {list(headers.keys())})")
-            # TODO: Refactor (Gohan)
-            # - Include table_id as part of the input parameters so Gohan can affiliate variants with a table
-            workflow_params["vcf_gz.table_id"] = table_id
+            try:
+                token = tt.get(scope, 1)
+            except Exception as err:
+                return flask_internal_server_error(err)
 
             # TODO: Refactor (Gohan)
             # - Pass TT in as a parameter to the workflow (used in the .wdl file directly)
             #    (current purpose for this is only to get automatic Gohan workflow requests through)
-            workflow_params["vcf_gz.temp_token"] = tr.json()[0]
-            workflow_params["vcf_gz.temp_token_host"] = urlparse(chord_url).netloc
+            workflow_params[f"{workflow_id}.temp_token"] = token[0]
+            workflow_params[f"{workflow_id}.temp_token_host"] = token_host
+
+            # TODO: Refactor (Gohan)
+            # - Include table_id as part of the input parameters so Gohan can affiliate variants with a table
+            workflow_params["vcf_gz.table_id"] = table_id
+
+        if "auth" in workflow_metadata:
+            for tok in workflow_metadata["auth"]:
+                tokenFactory = tt if tok.get("type", "tt") == "tt" else ott
```

Comment:
Python style consistency: `token_factory`; could add a type hint to indicate that it's an `AuthorizationToken` if you like

Reply:
Sure!